### PR TITLE
571 edit nodes continue

### LINKED ISSUE #571: Add possibility to edit and save modification in node view

### PR DESCRIPTION
**Pull request by kronken**
Closes #571

### CODE REVIEW

File: src/codeLists/nodeTypesCodeList.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+const nodeTypesCodeList = {
+    ['P']: 'Pysäkki',
+    ['X']: 'Risteys',
+    ['-']: 'Kuntaraja',
+    ['E']: 'Pois käytöstä',
+};
+
+export default nodeTypesCodeList;
```
Comment (Divinare) [inline src/codeLists/nodeTypesCodeList.ts]: nit: nodeTypeCodeList

**Comment by Divinare** [comment]
Pending until #600 is merged.

Implement isDirty with _.isEqual like in #600, remove hasUnsavedModifications variable

File: src/components/sidebar/networkView/NetworkView.tsx [hunk 2]
```diff
@@ -25,10 +25,6 @@ class NetworkView extends React.Component<INetworkViewProps> {
     }
 
     private initStores = () => {
-        this.props.networkStore!.selectAllTransitTypes();
-        this.props.networkStore!.showMapLayer(MapLayer.node);
-        this.props.networkStore!.showMapLayer(MapLayer.nodeWithoutLink);
-        this.props.networkStore!.showMapLayer(MapLayer.link);
         this.props.routeStore!.clearRoutes();
```
Comment (kronken) [inline src/components/sidebar/networkView/NetworkView.tsx]: YOLO

File: src/components/sidebar/networkView/nodeView/NetworkNode.tsx [hunk 3]
```diff
@@ -101,7 +150,12 @@ class NetworkNode extends React.Component<INetworkNodeProps, InetworkNodeState>
             <div className={s.networkNodeView}>
                 <div className={s.content}>
                     <ViewHeader
-                        closePromptMessage={undefined}
+                        closePromptMessage={
+                            this.props.nodeStore!.isDirty ? message : undefined
+                        }
+                        showEditButton={true}
+                        isEditing={!isEditingDisabled}
+                        onEditButtonClick={this.onEditButtonClick}
```
Comment (Divinare) [inline src/components/sidebar/networkView/nodeView/NetworkNode.tsx]: Rename onEditButtonClick = toggleIsEditingDisabled (everywhere else as well)

(name describing what function does is better than name a name of "what happened")

File: src/components/sidebar/networkView/nodeView/NetworkNode.tsx [hunk 4]
```diff
@@ -72,23 +79,65 @@ class NetworkNode extends React.Component<INetworkNodeProps, InetworkNodeState>
...
         const node = this.props.nodeStore!.node;
-        const isEditingDisabled = this.state.isEditDisabled;
+        const isEditingDisabled = this.state.isEditingDisabled;
+
+        const isSaveButtonDisabled = this.state.isEditingDisabled
+            || !this.props.nodeStore!.isDirty
+            || !this.isFormValid();
+
+        // tslint:disable-next-line:max-line-length
+        const message = 'Solmulla on tallentamattomia muutoksia. Oletko varma, että poistua näkymästä? Tallentamattomat muutokset kumotaan.';
```
Comment (Divinare) [inline src/components/sidebar/networkView/nodeView/NetworkNode.tsx]: const closePromptMessage = ...

**Comment by Divinare** [comment]
Remove validation of PITKÄ NIMI and PITKÄ NIMI RUOTSIKSI (they seem to be optional even though we will add/fix validations later)

File: src/components/sidebar/networkView/nodeView/StopForm.tsx [hunk 5]
```diff
@@ -93,9 +103,10 @@ const stopForm = ({ stop, onChange, isEditingDisabled }: IStopFormProps) => {
                         onChange={onChange('postalNumber')}
                     />
                     <Dropdown
-                        onChange={onChange('municipality')}
+                        onChange={onDropDownChange('municipality')}
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/components/sidebar/networkView/nodeView/StopForm.tsx]: Great that you noticed. This has been left there by mistake. Will remove

File: src/components/shared/inheritedComponents/FormBase.tsx [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    protected toggleIsEditingDisabled = (undoChange: () => void) => {
+        if (!this.state.isEditingDisabled) {
+            undoChange();
+        }
+        const isEditingDisabled = !this.state.isEditingDisabled;
+        this.setState({
+            isEditingDisabled,
+            invalidFieldsMap: {},
+        });
+    }
```
[the remark this answers is not in the document]
Comment (kronken) [reply src/components/shared/inheritedComponents/FormBase.tsx]: What if we consider this more in task: https://github.com/HSLdevcom/jore-map-ui/issues/612

**Comment by Divinare** [comment]
Still 1 bug exists:

1) Choose a node that is crossroad, for example: http://localhost:3000/network/node/4140028/?routes[0]=00332
2) Change "TYYPPI" to "Pysäkki"
3) Click "Edit button"
4) Click "Edit button" again
--> now "TYYPPI" should be "Risteys" but it is "Pysäkki"

* RoutePathViewForm should be RoutePathForm (we also have StopForm)

* NetworkNode should be renamed as NodeView


@ NetworkNode.tsx (consisten renaming):
onChange --> rename as: onNodeChange
onChangeCoordinates={this.onChangeLocations}
--> rename as:
onCoordinatesChange={this.onCoordinatesChange} (?)

**Pull request merged**